includes/Modules/Analytics/Advanced_Tracking.php

```diff
@@ -0,0 +1,141 @@
+<?php
+/**
+ * Class Google\Site_Kit\Modules\Analytics\Advanced_Tracking
+ *
+ * @package   Google\Site_Kit\Modules\Analytics
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Modules\Analytics;
+
+use Google\Site_Kit\Modules\Analytics\Advanced_Tracking\Plugin_Detector;
+use Google\Site_Kit\Modules\Analytics\Advanced_Tracking\Measurement_Event_Factory;
+use Google\Site_Kit\Modules\Analytics\Advanced_Tracking\Measurement_Code_Injector;
+
+// phpcs:disable WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+
+/**
+ * Class for Advanced Tracking.
+ *
+ * @since n.e.x.t.
+ * @access private
+ * @ignore
+ */
+final class Advanced_Tracking {
+
+	/**
+	 * List of plugins SiteKit supports for event tracking.
+	 *
+	 * @since n.e.x.t.
+	 * @var array of strings
+	 */
+	private $supported_plugins;
+
+	/**
+	 * List of event configurations to be tracked.
+	 *
+	 * @since n.e.x.t.
+	 * @var array
+	 */
+	private $event_configurations;
+
+	/**
+	 * Main class plugin detector instance.
+	 *
+	 * @since n.e.x.t.
+	 * @var Plugin_Detector
+	 */
+	private $plugin_detector;
+
+	/**
+	 * Advanced_Tracking constructor.
+	 *
+	 * @param Plugin_Detector $plugin_detector optional plugin detector used for testing.
+	 * @since n.e.x.t.
+	 */
+	public function __construct( $plugin_detector = null ) {
+		$this->supported_plugins = array(
+			array(
+				'name'       => 'Contact Form 7',
+				'check_name' => 'WPCF7_PLUGIN_DIR',
+				'check_type' => Plugin_Detector::$TYPE_CONSTANT,
+			),
+			array(
+				'name'       => 'Formidable Forms',
+				'check_name' => 'load_formidable_forms',
+				'check_type' => Plugin_Detector::$TYPE_FUNCTION,
+			),
+			array(
+				'name'       => 'Ninja Forms',
+				'check_name' => 'NF_PLUGIN_DIR',
+				'check_type' => Plugin_Detector::$TYPE_CONSTANT,
+			),
+			array(
+				'name'       => 'WooCommerce',
+				'check_name' => 'WC_PLUGIN_FILE',
+				'check_type' => Plugin_Detector::$TYPE_CONSTANT,
+			),
+			array(
+				'name'       => 'WPForms',
+				'check_name' => 'WPFORMS_PLUGIN_DIR',
+				'check_type' => Plugin_Detector::$TYPE_CONSTANT,
+			),
+			array(
+				'name'       => 'WPForms Lite',
+				'check_name' => 'WPFORMS_PLUGIN_DIR',
+				'check_type' => Plugin_Detector::$TYPE_CONSTANT,
+			),
+		);
+		$this->plugin_detector   = $plugin_detector;
+	}
+
+	/**
+	 * Registers functionality through WordPress hooks.
+	 *
+	 * @since n.e.x.t.
+	 */
+	public function register() {
+		add_action( 'wp_enqueue_scripts', array( $this, 'set_up_advanced_tracking' ), 11 );
+	}
```

Comment:
Something we haven't touched on yet, but talked about earlier, is that this PR also needs to support AMP - which to some degree is actually going to be easier because it only relies on a configuration object anyway (AMP does not support any custom JavaScript). Some initial pointers:
* Have a look add the [documentation for how to track `gtag` events with AMP](https://developers.google.com/analytics/devguides/collection/amp-analytics#measure_events).
* You can hook into Site Kit's `googlesitekit_amp_gtag_opt` filter which receives the entire `amp-analytics` configuration array. You can then alter its `triggers` value with the necessary event configurations data. Make sure to use the correct variable names (`event_name`, `event_category` etc).
    * We'd probably need some sort of utility method to transform an event configuration into the format that the `amp-analytics` config object expects.